presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -1427,10 +1431,12 @@ private ThriftMetastoreClient createMetastoreClient(HiveIdentity identity)
             return createMetastoreClient();
         }
 
-        String username = identity.getUsername()
-                .orElseThrow(() -> new IllegalStateException("End-user name should exist when metastore impersonation is enabled"));
-
         ThriftMetastoreClient client = createMetastoreClient();
+        String username = identity.getUsername().orElseThrow(() -> new IllegalStateException("End-user name should exist when metastore impersonation is enabled"));
+        if (authenticationType == HiveMetastoreAuthenticationType.KERBEROS) {
+            String delegationToken = getDelegationTokenOrClose(client, username);
```

Comment:
You don't need `getDelegationTokenOrClose` method. It would be simpler to write this like this

```java
switch (authenticationType) {
case KERBEROS:
    String delegationToken;
    try (ThriftMetastoreClient client = createMetastoreClient()) {
        delegationToken = client.getDelegationToken(username);
    }
    return clientProvider.createMetastoreClient(Optional.of(delegationToken));

case NONE:
    ThriftMetastoreClient client = createMetastoreClient();
    setMetastoreUserOrClose(client, username);
    return client;

default:
     throw new IllegalStateException("Unsupported authentication type: " + authenticationType);
}
```